### PR TITLE
[MRG] More work on the beamformer code

### PR DESCRIPTION
**Pull request by wmvanvliet**
This PR gives the `_create_beamformer` another refactoring pass to unify the LCMV and DICS flow. 

Bugs fixed:

 - LCMV beamformer when `pick_ori='normal', weight_norm='unit_noise_gain'`. The existing implementation has a bug where the weights were first normalized to have unit length *before* picking the normal direction. Instead, the weights should be normalized to unit length *after* picking the orientation, as is the case when `pick_ori='max-power'`.
 - DICS beamformer when `inversion='single', pick_ori='max-power', weight_norm='unit_noise_gain'`. The computation of the power (in order to pick the orientation with max power) was bugged.
 - DICS beamformer when `inversion='matrix', pick_ori='max-power', weight_norm='unit_noise_gain'`. In this case, a computational shortcut is performed, but it was performed incorrectly. See [this comment](https://github.com/mne-tools/mne-python/pull/5135/files#r200652209).

New features:

 - New fancy `reg_pinv` function, which now lives in `utils.py` and will even walk your dog.
 - No more branching between `method='lcmv'` and `method='dics'`.
 - The sign of the max power orientation is now always aligned with the sign of the surface normal (tnx @larsoner)

New features we got for free in the process:

 - `pick_ori='max-power'` now also supported when `weight_norm=None`
 - `weight_norm='nai'` now also supported when `pick_ori=None` and `pick_ori='normal'`
 - `weight_norm='nai'` now also supported for the DICS beamformer
 - `rank` parameter for the DICS beamformer to explicitly specify rank
 - `rank='auto'` setting to do rank-based inversion as in #5362 (DICS already did this)

In all other cases, the new version gives exactly the same output as the version in the current master branch:

### LCMV
pick_ori    | weight_norm       | new == old?
------------|-------------------|------------
None        | None              | True
None        | 'unit-noise-gain' | True
None        | 'nai'             | Not available in old
'normal'    | None              | True
'normal'    | 'unit-noise-gain' | True
'normal'    | 'nai'             | Not available in old
'max-power' | None              | Not available in old
'max-power' | 'unit-noise-gain' | True
'max-power' | 'nai'             | True

### DICS
pick_ori    | weight_norm       | inversion | new == old?
------------|-------------------|-----------|------------
None | None | 'single' | True
None | None | 'matrix' | True
None | 'unit-noise-gain' | 'single' | True
None | 'unit-noise-gain' | 'matrix' | True
None | 'nai' | 'single' | Not available in old
None | 'nai' | 'matrix' | Not available in old
'normal' | None | 'single' | True
'normal' | None | 'matrix' | True
'normal' | 'unit-noise-gain' | 'single' | True
'normal' | 'unit-noise-gain' | 'matrix' | True
'normal' | 'nai' | 'single' | Not available in old
'normal' | 'nai' | 'matrix' | Not available in old
'max-power' | None | 'single' | True
'max-power' | None | 'matrix' | True
'max-power' | 'unit-noise-gain' | 'single' | **False**
'max-power' | 'unit-noise-gain' | 'matrix' | **False**
'max-power' | 'nai' | 'single' | Not available in old
'max-power' | 'nai' | 'matrix' | Not available in old

### CODE REVIEW

**Comment by larsoner** [comment]
@wmvanvliet can you confirm that we should indeed be emitting warnings when using DICS with multiple channel types, e.g. mag+grad, or worse mag+grad+eeg?

**Comment by wmvanvliet** [comment]
for now, yes, but we should just whiten the CSD like we do with the LCMV beamformer. EIther me or @britta-wstnr will probably get to it soon.

**Comment by larsoner** [comment]
`tf_dics` failure (sometimes):

https://travis-ci.org/mne-tools/mne-python/jobs/417213324#L2690
https://travis-ci.org/mne-tools/mne-python/jobs/417213328#L2724

**Comment by wmvanvliet** [comment]
goody, that looks like fun to debug...

**Comment by wmvanvliet** [comment]
CircleCI is failing because sphynx-gallery refuses to build or something.

**Comment by wmvanvliet** [comment]
@britta-wstnr are you happy with the `weight_norm='nai'` as it's implemented in this PR?

**Comment by wmvanvliet** [comment]
Ok, tests all pass and this PR is ready for comments. @britta-wstnr WDYT?

File: mne/beamformer/_dics.py [hunk 1]
```diff
@@ -62,6 +63,11 @@ def make_dics(info, forward, csd, reg=0.05, label=None, pick_ori=None,
                 filters are computer for the orientation that maximizes
                 spectral power.
 
+    rank : int | 'auto' | None
+        The effective rank of the covariance matrix. If 'auto', the rank will
+        be estimated before regularization is applied. If ``None``, the rank
+        will be estimated after regularization is applied.
+        Defaults to 'auto'.
```
Comment (larsoner) [inline mne/beamformer/_dics.py]:
```
    Defaults to 'auto'.

    .. versionadded:: 0.17
inversion:
```

File: mne/beamformer/_dics.py [hunk 2]
```diff
@@ -165,7 +170,10 @@ def make_dics(info, forward, csd, reg=0.05, label=None, pick_ori=None,
            in Large-Scale Cortical Networks Predicts Perception.
            Neuron (2011) vol 69 pp. 387-396.
            https://doi.org/10.1016/j.neuron.2010.12.027
-    .. [4] van Vliet, et al. (2018) Analysis of functional connectivity and
+    .. [4] Van Veen et al. Localization of brain electrical activity via
+           linearly constrained minimum variance spatial filtering.
+           Biomedical Engineering (1997) vol. 44 (9) pp. 867--880
```
Comment (larsoner) [inline mne/beamformer/_dics.py]: Ahh nice. Does this mean you have checked that these changes don't break your results?

Comment (wmvanvliet) [reply mne/beamformer/_dics.py]: Yes. I've explicitly checked that the output of both `make_dics` and `make_lcmv` is identical to current master, as this is supposed to be just a refactoring. However, some bugs where exposed with some parameter combinations (see description of the PR at the very top), so the output in these cases is different.

Comment (larsoner) [reply mne/beamformer/_dics.py]: Excellent

**Comment by britta-wstnr** [comment]
@wmvanvliet I was hoping to find some time while at Biomag - but realized that I need some more time to go over the code! So excuse my slowness here, I will give it a look when back!

**Comment by wmvanvliet** [comment]
Enjoy Biomag!

**Comment by britta-wstnr** [comment]
Hi @wmvanvliet - finally had time looking over the code. However, it is basically impossible to compare old and new code snippets for the things that change the output of LCMV or DICS due to the refactoring. Could you please split the PR into 2, probably best code changes first and then refactoring?

**Comment by agramfort** [comment]
@wmvanvliet diff is indeed quite big and it is harder to review if part is refactoring and part is bug fix / change of behavior. Can you allocate some time to help @britta-wstnr review this work?

If you have doubts let me share a quote. "Along you go fast, together we go far" !

**Comment by wmvanvliet** [comment]
@britta-wstnr here are the changes to the LCMV beamformer, in a separate PR: #5511

**Comment by wmvanvliet** [comment]
Rebased after merging #5511

**Comment by larsoner** [comment]
@wmvanvliet in the meantime can you see if you agree https://github.com/wmvanvliet/mne-python/pull/6 is correct?

**Comment by wmvanvliet** [comment]
I will take a look!

**Comment by larsoner** [comment]
@britta-wstnr this would be great to get into 0.17. Can you take a look in the next week or so?

**Comment by larsoner** [comment]
@wmvanvliet does this close #5362?

**Comment by larsoner** [comment]
... and also @wmvanvliet we have to figure out how to deal with `whats_new.rst`. Perhaps in the entry you should add a URL link to this PR since you have a full description of the changes. Something like:

```
API
---
- Modified and unified DICS and LCMV computations, ... . A table listing the backward-incompatible changes from 0.16 to 0.17 can be viewed `here <https://github.com/mne-tools/mne-python/pull/5447>`__. By `Marijn van Vliet`_

**Comment by wmvanvliet** [comment]
> @wmvanvliet does this close #5362?

That's up to @britta-wstnr, it's her idea. Although I do think setting `reg=0, rank='auto'` comes very close to what she is proposing.

**Comment by agramfort** [comment]
@britta-wstnr do you have time to look? is this PR now smaller?

**Comment by wmvanvliet** [comment]
Rebased.

File: mne/beamformer/tests/test_lcmv.py [hunk 3]
```diff
@@ -330,21 +334,6 @@ def test_make_lcmv(tmpdir):
...
-                  pick_ori='normal', weight_norm='nai')
-    pytest.raises(NotImplementedError, make_lcmv, evoked.info, forward_vol,
-                  data_cov, reg=0.01, noise_cov=noise_cov, pick_ori=None,
-                  weight_norm='nai')
-
-    # Test if no weight-normalization and max-power source orientation throws
-    # an error
-    pytest.raises(NotImplementedError, make_lcmv, evoked.info, forward_vol,
-                  data_cov, reg=0.01, noise_cov=noise_cov,
-                  pick_ori='max-power', weight_norm=None)
```
Comment (larsoner) [inline mne/beamformer/tests/test_lcmv.py]: Hooray removing `NotImplementedError`!

Do you test running these options elsewhere? If not, then these should at least be turned into smoke tests (i.e., run the code but no need to check the output, just that it at least runs)

**Comment by britta-wstnr** [comment]
Hi guys, just back at the desk after some time off - gonna catch up over the next few days!

File: mne/beamformer/_compute_beamformer.py [hunk 4]
```diff
@@ -252,51 +207,145 @@ def _prepare_beamformer_input(info, forward, label, picks, pick_ori,
...
-                        pick_ori, reduce_rank, rank, is_free_ori,
-                        inversion=None):
-    """Compute a spatial filter (LCMV or DICS)."""
-    # Tikhonov regularization using reg parameter d to control for
-    # trade-off between spatial resolution and noise sensitivity
-    if method == 'lcmv':
-        Cm_inv, d = _reg_pinv(Cm.copy(), reg)
-    elif method == 'dics':
-        Cm_inv, _ = _reg_pinv(Cm, reg, rcond='auto')
+def _shortcut(Wk, Gk, Cm_inv_sq, reduce_rank, nn):
```
Comment (britta-wstnr) [inline mne/beamformer/_compute_beamformer.py]: Maybe use a more descriptive name, something like ``_norm_weights`` or such? 
Is it planned to use the following code more often? So far this function is only used once and it might be easier for future looks at the code to not have to jump back and forth?

File: mne/utils.py [hunk 5]
```diff
@@ -583,6 +605,103 @@ def _compute_row_norms(data):
     return norms
 
 
+def reg_pinv(x, reg=0, rank=None, rcond=1e-15):
+    """Compute a regularized pseudoinverse of a square matrix.
```
Comment (britta-wstnr) [inline mne/utils.py]: Is it right that this code not only moved but also changed a lot? If so, can you summarize the most important changes quickly? The ``diff`` is not of much help here ;-)

**Comment by britta-wstnr** [comment]
@larsoner and @wmvanvliet I think that #5362 has some differences to what is implemented here, especially in the way to estimate the rank - where we are also not done with thinking yet :smile: so I would vote to leave it open for now.

**Comment by agramfort** [comment]
should we plan a quick hangout on thursday to discuss this together?

**Comment by larsoner** [comment]
> I was wondering whether we should focus on shorter PRs for the future in the beamformer module, by the way. The code is quite new to all of us and large changes including both refactoring and changes to the implementation can be really hard to review - at least for me. What does the rest of you think?

Refactoring like this is almost always difficult to review. I haven't found a good way around it. If it's possible to split this into a first refactoring PR that does not change the output and a second one that does, it's possible it would help. On the other hand that first PR could still hide bugs easily, and the second probably wouldn't look a ton simpler than this PR, so I'm not sure the gain would be great.

But this depends on the particular PR, so we'll see what @wmvanvliet thinks.

Also in theory the unit tests (and example manual inspection) should be sufficient to make sure everything still works, but unfortunately this is more like a platonic ideal than reality :)

File: mne/utils.py [hunk 6]
```diff
@@ -563,17 +563,39 @@ def estimate_rank(data, tol='auto', return_singular=False, norm=True):
...
         data /= norms[:, np.newaxis]
     s = linalg.svd(data, compute_uv=False, overwrite_a=True)
+    rank = _estimate_rank_from_s(s, tol)
+    if return_singular is True:
+        return rank, s
+    else:
+        return rank
+
+
+def _estimate_rank_from_s(s, tol='auto'):
```
Comment (massich) [inline mne/utils.py]: can we use `_estimate_rank_from_singular_values`? or is `s` a well-known shorthand for `singular_values` in svd decomposition (I've no idea).

Comment (larsoner) [reply mne/utils.py]: s is standard

File: mne/beamformer/_compute_beamformer.py [hunk 7]
```diff
@@ -123,7 +67,11 @@ def _compare_ch_names(names1, names2, bads):
 
 def _check_one_ch_type(info, picks, noise_cov, method):
     """Check number of sensor types and presence of noise covariance matrix."""
+    # XXX : ugly hack to avoid picking subset of info with applied comps
```
[the remark this answers is not in the document]
Comment (larsoner) [reply mne/beamformer/_compute_beamformer.py]: This will need to be addressed more fully by Luke's open PR so I wouldn't bother changing this further

**Comment by larsoner** [comment]
FYI I pushed a couple commits to 1) remove now-unnecessary `info['comps']` hack and 2) rename `_shortcut` to something more informative.

File: mne/utils.py (hunk 5 above)
Comment (wmvanvliet) [reply mne/utils.py]: In the beamformer code in the current master branch, we use three methods to invert a matrix: `_reg_pinv`, `_eig_inv` and `scipy.linalg.pinv`. These three functions actually do something very similar: they invert a matrix through SVD- (or Eigen-) decomposition and only using *n* components in the inverse (=rank). Even `scipy.linalg.pinv` does nothing more than that. Where the functions differ is in the method of determining how many components to use.

So I thought it best to refactor `_reg_pinv` to a function that makes this explicit. It offers precise control over the number of components to use and thereby fulfils all the needs we have without having different inverse functions all over the place.

Importantly it offers these ways to set the rank (=number of components to use in inverse):

 1. Set it explicitly
 2. Use a threshold for the singular values *before* adding the regularization
 3. Use a threshold for the singular values *after* adding the regularization

Comment (larsoner) [reply mne/utils.py]: To check my understanding, (3) will produce a full rank output for any reg value not closer to zero than the threshold (i.e., not tiny), right?

Comment (larsoner) [reply mne/utils.py]: Cool. I'd be interested to hear if you have any thoughts about https://github.com/mne-tools/mne-python/pull/5481 since it seems closely related

**Comment by larsoner** [comment]
@britta-wstnr okay to merge this one, or is there more that we can/should do to move it forward? Are there consistency tests you want to run at your end?

File: mne/utils.py (hunk 5 above)
Comment (britta-wstnr) [reply mne/utils.py]: Aha, my bad. I think I got the notion of changes from ``reg_pinv`` being listed in New Features above ;-)

**Comment by britta-wstnr** [comment]
I think this is good to go. 
In general, I think it might be nice to get an idea how the DICS flavors implemented here compare to the DICS implementation in FieldTrip as I did for the LCMV @sarangnemo @agramfort

**Comment by larsoner** [comment]
Okay let's get this in, then. I agree a DICS comparison would be great moving forward

**Comment by larsoner** [comment]
@wmvanvliet can you add a suitable `whats_new.rst` entry before merge?

File: mne/beamformer/_lcmv.py [hunk 8]
```diff
@@ -437,7 +441,7 @@ def _lcmv_source_power(info, forward, noise_cov, data_cov, reg=0.05,
     # Tikhonov regularization using reg parameter to control for
     # trade-off between spatial resolution and noise sensitivity
     # This modifies Cm inplace, regularizing it
-    Cm_inv, d = _reg_pinv(Cm, reg)
+    Cm_inv, d, _ = reg_pinv(Cm, reg)
```
Comment (larsoner) [inline mne/beamformer/_lcmv.py]: Also let's keep `reg_pinv` private for now (I can do this, too, if I'm going to update the docstrings)

**Comment by agramfort** [comment]
thanks @larsoner for the wrap up.

small steps for rank estimation big steps for ...

**Comment by larsoner** [comment]
So @agramfort you agree the two steps I mention above (add note to docstring, fully unify in 0.18) would be sufficient?

File: mne/beamformer/_lcmv.py (hunk 8 above)
Comment (wmvanvliet) [reply mne/beamformer/_lcmv.py]: Good idea. Feel free to do this yourself :)

**Comment by wmvanvliet** [comment]
It would be great it we could merge the code for rank estimation across `compute_whitener` and `reg_pinv`. And even better if the API would remain backwards compatible. That gives reason to think a bit about the current behavior of the `rank` parameter in both functions.

Right now, `rank=None` actually performs automatic rank detection for both functions. In the case of `make_lcmv` and `make_dics`, it is deferred to `scipy.linalg.pinv` to do this. Maybe we should rename the `rank='auto'` parameter to something else. Then we can merge the rank estimation code and retain the current API.

**Comment by wmvanvliet** [comment]
Oh and no, as I explained to @britta-wstnr, the estimation code has **not** changed. The result is exactly the same as before.

**Comment by larsoner** [comment]
> Right now, rank=None actually performs automatic rank detection for both functions. In the case of make_lcmv and make_dics, it is deferred to scipy.linalg.pinv to do this. Maybe we should rename the rank='auto' parameter to something else. Then we can merge the rank estimation code and retain the current API.

How about we have it so that:

1. `rank=None` means "estimate rank, and preserve it during regularization", which I think is what `rank='auto'` does now. That would be consistent with what minimum-norm `rank=None` does (preserves rank) and also thus what I'm having it do in #5481.
2. `rank=False` means "regularize then compute rank". After regularization (unless the reg param is 0) this is always full rank, thus why "False" is reasonable here.

**Comment by wmvanvliet** [comment]
Yes, that's a good idea! Feel free to make it so @larsoner!

File: mne/beamformer/_lcmv.py [hunk 9]
```diff
@@ -441,7 +442,7 @@ def _lcmv_source_power(info, forward, noise_cov, data_cov, reg=0.05,
     # Tikhonov regularization using reg parameter to control for
     # trade-off between spatial resolution and noise sensitivity
     # This modifies Cm inplace, regularizing it
-    Cm_inv, d, _ = reg_pinv(Cm, reg)
+    Cm_inv, d, _ = _reg_pinv(Cm, reg, rank=rank)
```
Comment (larsoner) [inline mne/beamformer/_lcmv.py]: @wmvanvliet I think you just missed this one as every other `Cm_inv, ..., _reg_pinv` preceded by a Tikhonov comment has `rank=rank`, but maybe I'm missing a subtlety here

File: mne/beamformer/_lcmv.py [hunk 10]
```diff
@@ -423,6 +422,8 @@ def _lcmv_source_power(info, forward, noise_cov, data_cov, reg=0.05,
         info, [info['ch_names'].index(k) for k in ch_names
                if k in info['ch_names']])
 
+    # XXX this could maybe use pca=True to avoid needing to use
+    # _reg_pinv(..., rank=rank) later
```
Comment (larsoner) [inline mne/beamformer/_lcmv.py]: Just a minor comment here that eventually we could make `compute_beamformer` pass on the `pca` argument (we currently don't publicly expose it), meaning that it would return a non-rank-deficient multplier of shape `(rank, 306)` instead of `(306, 306)`. It might simplify things at some point. I've also wanted this parameter before when analyzing data.

**Comment by larsoner** [comment]
Okay @wmvanvliet done, feel free to look to make sure it looks like I did not break anything. I touched the beamformer examples so we could ensure that they are not broken:

- [PR](https://10082-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_lcmv_beamformer.html) - [master](http://mne-tools.github.io/dev/auto_examples/inverse/plot_lcmv_beamformer.html)
- [PR](https://10082-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_lcmv_beamformer_volume.html) - [master](http://mne-tools.github.io/dev/auto_examples/inverse/plot_lcmv_beamformer_volume.html)
- [PR](https://10082-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_dics.html) - [master](http://mne-tools.github.io/dev/auto_tutorials/plot_dics.html)

@wmvanvliet it looks like the DICS example changed, maybe not for the better. It looks like it's also that way on your commit when I try a8f4f1fe7c1da66b002c5845c7a9664098e588f7 locally. Do you think it's okay, or is there something to fix?

**Comment by wmvanvliet** [comment]
I'll have to take a look. Don't merge this yet. Currently busy with kids but I'll try to get to it soon.

File: mne/beamformer/_compute_beamformer.py [hunk 11]
```diff
@@ -301,10 +301,11 @@ def _compute_beamformer(G, Cm, reg, n_orient, weight_norm, pick_ori,
         The source orientation to compute the beamformer in.
     reduce_rank : bool
         Whether to reduce the rank by one during computation of the filter.
-    rank : int | 'auto' | None
-        The effective rank of the covariance matrix. If 'auto', it will be
-        estimated before regularization is applied. If ``None``, it will be
-        estimated after regularization is applied.
+    rank : None | False | int
```
Comment (wmvanvliet) [inline mne/beamformer/_compute_beamformer.py]: On second thought, `None`, `False`, `int` is very unintuitive. Also, it should default to `False` to keep the current behavior (this is probably why the example figure has changed)

File: mne/beamformer/_dics.py [hunk 12]
```diff
@@ -63,11 +63,11 @@ def make_dics(info, forward, csd, reg=0.05, label=None, pick_ori=None,
...
-    rank : int | 'auto' | None
-        The effective rank of the covariance matrix. If 'auto', the rank will
-        be estimated before regularization is applied. If ``None``, the rank
-        will be estimated after regularization is applied.
-        Defaults to 'auto'.
+    rank : None | False | int
+        The effective rank of the covariance matrix.
+        If None, the rank will be estimated before regularization is
+        applied. If False, the rank will be estimated after regularization
+        is applied. Defaults to ``None``.
```
Comment (wmvanvliet) [inline mne/beamformer/_dics.py]: Yes, here it should default to `None` and for the LCMV it should default to `False`

File: mne/beamformer/_lcmv.py (hunk 9 above)
Comment (wmvanvliet) [reply mne/beamformer/_lcmv.py]: Ah yes, I missed this one.

**Comment by wmvanvliet** [comment]
@larsoner how about this? `rank='full'` makes more sense to me than `rank=False`.

**Comment by larsoner** [comment]
Yeah that makes sense. Travis is red, though, can you look into it? If not I can do it in a few hours

**Comment by larsoner** [comment]
@wmvanvliet so for `lcmv` it looks like `rank=None` was being passed to `compute_whitener` before by default. I have made it so that `rank='full'` still does this to maintain backward-compatible behavior, even though it's a bit confusing. `compute_whitener` still gives an array of full size `(n_ch, n_ch)`, though, so in the end the `reg` will still make this full rank, so I think it's okay.

File: doc/whats_new.rst [hunk 13]
```diff
@@ -165,10 +172,15 @@ Bug
...
 - Fix error in mne coreg when saving with scaled MRI if fiducials haven't been saved by `Ezequiel Mikulan`_
 
+- Fix normalization error in :func:`mne.beamformer.make_lcmv` when ``pick_ori='normal', weight_norm='unit_noise_gain'`` by `Marijn van Vliet`_
+
+- Fix computation of max-power orientation in :func:`mne.beamformer.make_dics` when ``pick_ori='max-power', weight_norm='unit_noise_gain'`` by `Marijn van Vliet`_
 
 API
 ~~~
 
+- The parameter ``rank`` was added to :func:`mne.beamformer.make_dics` and :func:`mne.beamformer.make_lcmv`` with a default of ``'full'`` that will change to ``None`` in 0.18 to auto-compute the rank of the covariance matrix before regularization by `Marijn van Vliet`_
```
Comment (larsoner) [inline doc/whats_new.rst]: @wmvanvliet I realized that we should probably do this. We need a deprecation cycle for `make_lcmv` to do this. I assume we also need it for `make_dics` but if not let me know and I can set the defaults there back to `None` instead of `''` (the internal signal I used for checking deprecation).

File: mne/beamformer/_dics.py (hunk 12 above)
Comment (larsoner) [reply mne/beamformer/_dics.py]: Just re-read this and realized that this means there should be no deprecation cycle for DICS. I'll revert the deprecation there

**Comment by larsoner** [comment]
Updated examples above to reflect their current status. @wmvanvliet if you are happy with the deprecation updates I made and the examples, then I think this is finally ready for merge!

File: mne/beamformer/tests/test_lcmv.py [hunk 14]
```diff
@@ -276,7 +276,7 @@ def test_make_lcmv(tmpdir):
                                            bem=sphere, eeg=False, meg=True)
 
     # Test that we get an error if not reducing rank
-    with pytest.raises(ValueError, match='Singular matrix'):
+    with pytest.raises(ValueError):  # Singular matrix or complex spectrum
```
Comment (wmvanvliet) [inline mne/beamformer/tests/test_lcmv.py]: Why can't we explicitly test for error content?

File: doc/whats_new.rst (hunk 13 above)
Comment (wmvanvliet) [reply doc/whats_new.rst]: The default for `make_dics` is already `None`. For the LCMV, I don't like changing the default without some good simulation study backing it up. Both the LCMV and DICS defaults are carefully selected to yield generally stable beamformers.

Comment (wmvanvliet) [reply doc/whats_new.rst]: @britta-wstnr care to weight in on this?

**Comment by larsoner** [comment]
Okay reverted the deprecation. It seems like it *should* be okay to have it in there but I agree it's safer not to. And we want to release ASAP, so let's not do the deprecation "full"->None in 0.17.

@wmvanvliet do you want to try to fix the "Singular matrix" vs "complex spectrum" error depending on build type? To me it seems not worth the effort, unless you see that it's obvious and easy, e.g. some default tolerance is bit too high and fixing it makes it work.

**Comment by wmvanvliet** [comment]
Hang on... something changed in the output of the `plot_lcmv_beamformer.py` example. This shouldn't happen, as the max-power LCMV beamformer should not have changed at all (see table at the top). I'll need to see what exactly caused the change, it may be a bug. Sorry for yet another delay :)

**Comment by larsoner** [comment]
I did change the example to:

1. plot the signed values
2. use the `abs(max(...))` instead of `max(...)` to find the STC max value to plot
3. pass `rank=None`, but this only changes the estimate changes a tiny bit

It looks okay to me with these differences in mind. (It actually looks better, since (2) makes it so the same vertex is now found in free orientation as in fixed.)

**Comment by larsoner** [comment]
I can confirm that this (reverting) diff:
```
-    max_idx = np.argmax(np.abs(stc.data[:, time_idx]))
+    max_idx = np.argmax(stc.data[:, time_idx])
... 
-brain = stc.plot(hemi='lh', views='lat', subjects_dir=subjects_dir,
-                 initial_time=0.1, time_unit='s', smoothing_steps=5)
+brain = abs(stc).plot(hemi='lh', views='lat', subjects_dir=subjects_dir,
+                      initial_time=0.1, time_unit='s')
```
Puts the `plot_lcmv_beamformer.py` brain image back as:

screen shot 2018-11-04 at 19 27 49

But I think the one in this PR is more correct (using `max(abs(...))`; all three `max(abs(...))` vertices are the same!) and more informative (plotting signed data so you can see the signed-ness).

**Comment by wmvanvliet** [comment]
ah ok. Thanks for looking into it.

**Comment by larsoner** [comment]
Thanks @wmvanvliet !

**Comment by mmagnuski** [comment]
🎈 🍾 🎉
